Define representability, universal morphisms, and their isomorphism

maxsnew opened the pull request:
A fundamental theorem in category theory, for instance Proposition 2.4.8 in https://math.jhu.edu/~eriehl/context.pdf . Mostly follows as a corollary of Yoneda.

In writing it I added a few useful things to the presheaf, Yoneda and category of elements modules.

If we prove that the category of elements of P is univalent when C is univalent then we would also get that the type of representations is a prop. Should be a mild generalization of the proof for slice categories.

maxsnew commented:
I'm converting this to draft as I have developed a lot more code along the same lines that should be useful to bundle together.

maxsnew commented:
Updated the description above. Ready for review now

felixwellen commented:
Thanks for all this stuff, this looks really useful!

felixwellen commented:
We don't have a strict 80-line policy, but we try to avoid very long lines. So use your own judgement, but I think in the following files, some lines are too long:

Cubical/Categories/Constructions/Free/Category/Base.agda
Cubical/Categories/Constructions/Free/Functor.agda
Cubical/Categories/Yoneda.agda

felixwellen commented:
Cubical/Tactics/Reflection.agda

Do you think I could also use that to refactor the solver for CommRings/Monoids/...?

maxsnew commented:
> Cubical/Tactics/Reflection.agda
> 
> Do you think I could also use that to refactor the solver for CommRings/Monoids/...?

I think so. Even better if it were in agda-base as it's modified 1lab code already.

maxsnew commented:
I'm working on the column overflows (sorry). But I did push one actual (minor) change to the Free category implementation that I had already needed downstream.

maxsnew commented:
Ok column overflows should be good now.

felixwellen commented on Cubical/Tactics/FunctorSolver/Solver.agda:
```diff
@@ -0,0 +1,64 @@
+{-# OPTIONS --safe --lossy-unification #-}
```
Credit the 1lab?

maxsnew replied:
The implementations of the solvers are not taken from the 1lab, just the core reflection stuff in `Cubical.Tactics.Reflection`

felixwellen commented on Cubical/Categories/UnderlyingGraph.agda:
```diff
@@ -0,0 +1,54 @@
+
+-- Underlying graph of a category
+Ugr : ∀ {ℓc ℓc'} (𝓒 : Category ℓc ℓc') → Graph ℓc ℓc'
+Ugr 𝓒 .Node = 𝓒 .ob
```
I would write that one out (at least the graph part)

felixwellen commented on Cubical/Categories/UnderlyingGraph.agda:
```diff
@@ -0,0 +1,54 @@
+
+module Cubical.Categories.UnderlyingGraph where
+
+open import Cubical.Categories.Morphism
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.HLevels
+open import Cubical.Categories.Category.Base
+open import Cubical.Categories.Functor
+open import Cubical.Categories.NaturalTransformation hiding (_⟦_⟧)
+open import Cubical.Data.Graph.Base
+open import Cubical.Data.Sigma
+open import Cubical.Categories.Isomorphism
```
We use the (undocumented - sorry) convention to start with the most basic imports like "Foundations" and group imports. In this case:
```
open import Cubical.Foundations.Prelude
open import Cubical.Foundations.HLevels

open import Cubical.Data.Graph.Base
open import Cubical.Data.Sigma

open import Cubical.Categories.Category.Base
open import Cubical.Categories.Functor
open import Cubical.Categories.Morphism
open import Cubical.Categories.NaturalTransformation hiding (_⟦_⟧)
open import Cubical.Categories.Isomorphism
```

maxsnew replied:
Is there some particular order? Like Foundations then Data/HITs then Categories?

felixwellen replied:
There is no fixed standard, but most of us use the lose convention that things get more specific from top to bottom. Which means "Foundations then Data/HITs then Categories" in this case.

felixwellen commented on Cubical/Categories/Presheaf/Morphism.agda:
```diff
@@ -0,0 +1,90 @@
+{-# OPTIONS --safe #-}
+module Cubical.Categories.Presheaf.Morphism where
```
I think "Morphism" is a bit confusing terminology - is it standard? If yes, please state that. Reading below, I would have expected a name like "Heteromorphism" or something like that.

maxsnew replied:
If you view the presheaves as fibrations, then this is the usual notion of "morphism of fibrations" e.g. (https://ncatlab.org/nlab/show/Grothendieck+fibration) so I think it's a reasonable name.

felixwellen replied:
Ah, yes ;-)
Maybe that should be at the start of the comment below and that comment should be at the start of the file.

felixwellen commented on Cubical/Categories/Constructions/Free/Functor.agda:
```diff
@@ -0,0 +1,346 @@
+              (λ i k → aom f₁ k)
+              i j k
+
+        -- TODO
+        -- sem-uniq-ϕ : Square arb𝓕
+        --                     (idToPath semϕ)
+        --                     (λ i → 𝓕 ∘F sem-uniq-G i)
+        --                     (λ i → sem-uniq-H i ∘F Fϕ)
+        -- sem-uniq-ϕ = {!!}
+
+        -- TODO: uniqueness of the uniqueness paths above
```
It would be good to give more context, if this TODO is left. I think, in general it is better to give a summary comment at the start of the file, what has been done and what is still missing.

felixwellen commented on Cubical/Categories/Instances/Functors.agda:
```diff
@@ -102,3 +103,59 @@ module _ (C : Category ℓC ℓC') (D : Category ℓD ℓD') where
+          ≡⟨ F .F-seq (f , C .id) (g , C .id) ⟩
+        (F ⟪ g , C .id ⟫) ∘⟨ D ⟩ (F ⟪ f , C .id ⟫) ∎
+
+  -- TODO: UMP relating λF and appF
```
Maybe also here, the status could be described at the start of the file.

felixwellen reviewed (changes requested):
In general, the PR looks good and contains a lot of very useful stuff!
One important thing is, that we document what was taken from the 1lab (they use a different license and we should credit them properly in any case).
Otherwise, I mostly commented on things on the level of style questions and asked for more documentation.

maxsnew commented on Cubical/Categories/Presheaf/Base.agda:
```diff
@@ -22,3 +22,23 @@ PresheafCategory C ℓS = FUNCTOR (C ^op) (SET ℓS)
+action C P = P .F-hom
+
+-- Convenient notation for naturality
+syntax action C P f ϕ = C [ ϕ ∘ᴾ⟨ P ⟩ f ]
```
It comes from thinking of a presheaf as an *action* of a category on a collection of sets. So just like in a group action we can use the same notation for the action as the multiplication.

Though in retrospect, I think a better notation would be entirely infix like `ϕ ∘ᴾ⟨ C , P ⟩ f`

maxsnew replied:
Unfortunately the C can't be implicit.

felixwellen replied:
Yes, once we move to latest agda master, we can try that out. But I wouldn't try to do this in this PR.

jpoiret replied:
Either way, I don't think yet another notation for `P ⟪ f ⟫ ϕ` is a good idea. We can either keep 2.6.3 for now and develop everything for covariant presheaves, this way there's no problem with `C` being implicit and `P ⟪ f ⟫ ϕ` can be used, or wait for the switch to Agda master. If everyone is ok with 1, I can take care of it.

maxsnew commented on Cubical/Tactics/Reflection.agda:
```diff
@@ -0,0 +1,102 @@
+
+{- Utilities common to different reflection solvers.
+
+  Most of these are copied/adapted from the 1Lab
```
Just this file is from the 1lab. I can add a link to the file but let me know if there's a specific format I should follow.

felixwellen replied:
Yes, a link is good. And maybe "1Lab (AGPL 3.0)" to make it clear, they use a different license. I'm afraid I don't actually know how to do things like this properly.

jpoiret replied:
It would probably be good to follow the SPDX format for this, as it is quite standard now. You can have a look at https://spdx.dev/ids/ for how to do so!

jpoiret replied:
IANAL but looking at this closer, GPL is more restrictive than MIT and you can't license GPL code under MIT. I think cubical would have to also include the AGPL license, and be distributed under a dual license.

felixwellen replied:
Maybe adding something to the MIT-license file like: "Note that a more restrictive license might apply to  some individual files" is enough?

felixwellen replied:
This is now tracked as #1018. The solution there includes adding spdx identifiers, so please do that where necessary.

plt-amy replied:
I'm a tad too tired to think about doing a large-scale re-license this week, but I have no issue with treating the 1Lab.Reflection tree as BSD-3-Clause; the other contributors to those files are @ncfavier and @totbwf.

TOTBWF replied:
I have absolutely no issues with having this file be BSD-3 either

ncfavier replied:
Same

felixwellen replied:
Thanks everyone for resolving this!
So we use "BSD-3-Clause" for this file.

maxsnew commented:
Ok I've addressed the last changes with my most recent commits:
1. Add AGPL >=3.0 license to the file adapted from 1lab
2. fix import order
3. Update to work with the recent changes to `Data.Equality`/`Cubical.Foundations.Id`.
4. Refactor UniversalElement to be defined in terms of `isEquiv`. I also changed the names as @felixwellen suggested an put an extended comment in `Presheaf.Representable` to explain what the file does in more detail.

For the last point, the equivalence between terminal element and universal element would be simpler if the morphisms in the category of elements were defined differently as @jpoiret mentioned, but a non-trivial amount of code would have to be rewritten around the new definition and I didn't want to increase the scope of this PR any more so I didn't do that.

jpoiret commented:
So, I think the last missing bit is marking the 1lab file as BSD-3 instead of AGPLv3. Could you also remove the commented code in `Cubical.Categories.Presheaf.Representable`? Regarding the `Elements` change, I have a commit doing it, so I can send a separate PR after yours, it shouldn't be blocking :)

maxsnew commented:
Fixed the license and removed the accidentally included comments in the latest force push.

felixwellen reviewed (approved):
Should be good to merge now.
